contracts/services/reverse-splitter/src/contract.rs

```diff
@@ -0,0 +1,301 @@
+                // Lookup account/denom balance
+                let account = split.account();
+                let denom = split.denom();
+                let balance = account_balances.get(&account_key(account, denom)).unwrap();
+                let (amount, ratio, factor) = if let SplitAmount::FixedAmount(amount) =
+                    split.amount()
+                {
+                    (*amount, Decimal::one(), &None::<u64>)
+                } else {
+                    let ratio = match split.amount() {
+                        SplitAmount::FixedRatio(ratio) => *ratio,
+                        SplitAmount::DynamicRatio {
+                            contract_addr,
+                            params,
+                        } => *dynamic_ratios
+                            .get(&dyn_ratio_key(denom, contract_addr, params))
+                            .unwrap(),
+                        _ => unreachable!(),
+                    };
+                    let mut amount = balance.multiply_ratio(ratio.numerator(), ratio.denominator());
+                    if let Some(factor) = split.factor() {
+                        amount = amount
+                            .checked_div((*factor as u128).into())
+                            .map_err(|err| ServiceError::Std(err.into()))?;
+                    }
+                    (amount, ratio, split.factor())
+                };
+                Ok((amount, ratio, factor, denom, account))
+            })
+            .collect::<Result<Vec<_>, ServiceError>>()?;
+
+        // Find the minimum amount in base denom
+        let min_amount_in_base_denom = *amounts_in_base_denom
+            .iter()
+            .map(|(amount, _, _, _, _)| amount)
+            .min()
+            .unwrap();
+
+        amounts_in_base_denom
+            .into_iter()
+            .map(|(_, ratio, factor, denom, account)| {
+                // Divide min amount by ratio (invert ratio's numerator and denominator and multiply)
+                let mut amount =
+                    min_amount_in_base_denom.multiply_ratio(ratio.denominator(), ratio.numerator());
+                if let Some(factor) = factor {
+                    amount = amount
+                        .checked_mul((*factor as u128).into())
+                        .map_err(|err| ServiceError::Std(err.into()))?;
+                }
+                Ok((amount, denom, account))
+            })
+            .collect::<Result<Vec<_>, _>>()
+    }
+
+    fn query_dynamic_ratio(
+        querier: &QuerierWrapper<Empty>,
+        contract_addr: &Addr,
+        params: &str,
+        denom: &CheckedDenom,
+    ) -> Result<Decimal, ServiceError> {
+        let denom_name = denom.to_string();
+        let res: DynamicRatioResponse = querier.query_wasm_smart(
+            contract_addr,
+            &DynamicRatioQueryMsg::DynamicRatio {
+                denoms: vec![denom_name.clone()],
+                params: params.to_string(),
+            },
```

Comment:
Regarding querying a dynamic ration, we would probably need a wrapper of our own that will support this query because 99% other contracts wont.
Just something to note somewhere so we won't forgot.